DOC Fix tree explanation of `tree_.value` in example

adam2392 opened the pull request:
#### Reference Issues/PRs
Fixes: https://github.com/scikit-learn/scikit-learn/issues/28921

#### What does this implement/fix? Explain your changes.
Fixes the explanation of `tree_.value` as suggested in https://github.com/scikit-learn/scikit-learn/issues/28921#issuecomment-2174945248.

A follow-on PR that changes the name of `tree_.value` with a deprecation to `tree_.weighted_proportion` seems better and more explicit.

#### Any other comments?
cc: @lesteve

adam2392 committed:
Fix tree explanation

Signed-off-by: Adam Li <adam2392@gmail.com>

ArturoAmorQ reviewed (commented):
Hi @adam2392, thanks for addressing this issue :) Here's a couple of comments:

- On top of the API changelog entry, we might want to also include a `versionchanged` directive for each of the attributes on those estimators.

- For the FIX changelog entry, I think you are linking to an unrelated PR, but in any case, the `proportion` parameter of [`plot_tree`](https://scikit-learn.org/dev/modules/generated/sklearn.tree.plot_tree.html) function has not changed in version 1.6 as far as I know, or am I misunderstanding something?

adam2392 commented:
> * On top of the API changelog entry, we might want to also include a `versionchanged` directive for each of the attributes on those estimators.

The `tree_.value` is currently an undocumented attributed, so should I add a `versionchanged` to the `tree_` attribute in each of these tree classes?

> * For the FIX changelog entry, I think you are linking to an unrelated PR, but in any case, the `proportion` parameter of [`plot_tree`](https://scikit-learn.org/dev/modules/generated/sklearn.tree.plot_tree.html) function has not changed in version 1.6 as far as I know, or am I misunderstanding something?

The PR that I linked is when the behavior changed such that `tree_.value` in Cython stores proportions instead of absolute counts. The `plot_tree` does not have any API changes you're right.

adam2392 committed:
Address arturo changes

Signed-off-by: Adam Li <adam2392@gmail.com>

ArturoAmorQ committed:
Merge branch 'main' into treedescription

adam2392 commented:
The issues raised are fixed now:

1. only show floating point up to 3 decimal places when printing `tree_.value` in the example
2. show proportions when plotting tree (here I added an API change to set `proportion=True` by default, since this is in line with the behavior of `tree_.value`. (pending commit)

ref: https://output.circle-artifacts.com/output/job/bdc192fe-d4f9-4649-9ff0-03ddfb038abf/artifacts/0/doc/auto_examples/tree/plot_unveil_tree_structure.html

adam2392 committed:
Change default parameter

Signed-off-by: Adam Li <adam2392@gmail.com>

ArturoAmorQ commented:
I would rather not change the behavior of `plot_tree` for backward compatibility and just set `proportion=True` in the example. Could you please revert the latest commit?

adam2392 committed:
Revert

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 commented:
Done

ArturoAmorQ commented on sklearn/tree/_export.py:
```diff
@@ -150,7 +150,8 @@ def plot_tree(
     proportion : bool, default=False
         When set to ``True``, change the display of 'values' and/or 'samples'
-        to be proportions and percentages respectively.
+        to be proportions and percentages respectively. When set to ``False``,
+        change the display of 'values' and/or 'samples' to be counts.
```
This still has to be reverted.

ArturoAmorQ commented on examples/tree/plot_unveil_tree_structure.py:
```diff
@@ -148,6 +159,8 @@
+# Here, we show the proportions of samples of each class that reach each
+# node corresponding to the actual elements of `tree_.value` array.
 
 tree.plot_tree(clf)
```
You still have to set `proportion=True` to match the text.

adam2392 replied:
Ah right, I reverted it when I added the deprecation, but forgot to unrevert when I un-reverted the deprecation

ArturoAmorQ reviewed (approved):
A couple of final comments but otherwise LGTM :)

ArturoAmorQ committed:
Apply suggestions from code review

ArturoAmorQ commented on doc/whats_new/v1.4.rst:
```diff
@@ -937,6 +937,13 @@ Changelog
+  :class:`tree.DecisionTreeRegressor`, :class:`tree.ExtraTreeClassifier` and
+  :class:`tree.ExtraTreeRegressor` changed from an weighted absolute count
+  of number of samples to a weighted fraction of the total number of samples.
+  :pr:`13649` by :user:`Samuel Ronsin <samronsin>`, initiated by
+  :user:`Patrick O'Reilly <pat-oreilly>`.
```
Pinging @glemaitre for his opinion to rather linking #27639 here.

adam2392 replied:
Edited and moved to the top.

adam2392 committed:
Merge branch 'main' into treedescription

adam2392 committed:
Merge branch 'main' into treedescription

lesteve commented on sklearn/tree/_classes.py:
```diff
@@ -883,6 +883,9 @@ class DecisionTreeClassifier(ClassifierMixin, BaseDecisionTree):
         :ref:`sphx_glr_auto_examples_tree_plot_unveil_tree_structure.py`
         for basic usage of these attributes.
 
+        .. versionchanged:: 1.4
```
Not sure this makes sense to document a nested attributed change ...

The fact that nobody noticed/complained led me to believe that this is an implementation detail change that is not worth putting forward like this.

adam2392 replied:
Removed this then.

lesteve commented on examples/tree/plot_unveil_tree_structure.py:
```diff
@@ -122,24 +122,35 @@
 #
 # For example, in the above tree built on the iris dataset, the root node has
-# ``value = [37, 34, 41]``, indicating there are 37 samples
+# ``value = [0.331, 0.304, 0.366]`` indicating there are 33.1% of class 0 samples,
```
The printed output and `plot_tree` says  0.33 rather than 0.331 in the example doc.

Othere places need to be fixed similarly.

adam2392 replied:
Fixed this to say 0.33 now

lesteve reviewed (commented):
Thanks a lot, a few comments from me

adam2392 committed:
Update the comments from lesteve

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 commented:
Thanks for the review @lesteve! I have addressed your comments in [7343096](https://github.com/scikit-learn/scikit-learn/pull/29331/commits/7343096c0d6dacbd0cc347ca882881f0e63bc41f)

adam2392 committed:
Remove extra changes

Signed-off-by: Adam Li <adam2392@gmail.com>

lesteve committed:
Update v1.4.rst

lesteve commented on examples/tree/plot_unveil_tree_structure.py:
```diff
@@ -122,24 +123,35 @@
+#
+# One could convert this to the absolute weighted number of samples reaching a node,
+# by multiplying this number by `tree_.weighted_n_node_samples[node_idx]` for the
+# given node, or by `tree_.n_node_samples[node_idx]` for the unweighted number of
```
I am a bit unsure about multiplying `tree_.value` (proportion of samples taking into account sample weights) by `tree_.n_node_samples[node_idx]` which does not take sample weights into account. It seems to me that the output of this multiplication has not a clear meaning, but maybe I am missing something ...

Maybe only mention `tree_.weighted_n_nodes_sample` and mention that sample weights are not used in this example, which is the same as being one for all the samples?

adam2392 replied:
Done: [3c1f24a](https://github.com/scikit-learn/scikit-learn/pull/29331/commits/3c1f24a51e0d989ca261168716a208b4c5d186f6)

adam2392 commented on examples/tree/plot_unveil_tree_structure.py:
```diff
@@ -148,8 +160,10 @@
+# node corresponding to the actual elements of `tree_.value` array.
 
-tree.plot_tree(clf)
+tree.plot_tree(clf, proportion=True)
```
I originally did deprecation, but @ArturoAmorQ (https://github.com/scikit-learn/scikit-learn/pull/29331#issuecomment-2189090862) suggested I keep the default behavior.

adam2392 replied:
I honestly think either way is okay as this is like power-user stuff

adam2392 committed:
Address lesteve comment

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Merge branch 'main' into treedescription

lesteve commented:
OK thanks a lot for your patience and continued work on this @adam2392, let's merge this one!
